client/lib/plans/index.js

```diff
@@ -214,11 +203,111 @@ export function getYearlyPlanByMonthly( planSlug ) {
+	return getPlan( planSlug ).type === TYPE_BUSINESS;
+}
+
+export function isPremiumPlan( planSlug ) {
+	return getPlan( planSlug ).type === TYPE_PREMIUM;
+}
+
+export function isPersonalPlan( planSlug ) {
+	return getPlan( planSlug ).type === TYPE_PERSONAL;
+}
+
+export function isFreePlan( planSlug ) {
+	return getPlan( planSlug ).type === TYPE_FREE;
+}
+
+/**
+ * @see findSimilarPlansKeys
+ *
+ * @param {string|object} planKey Source plan to compare to
+ * @param {object} diff Properties that should differ in matched plan. @see planMatches
+ * @return {string|undefined} Matched plan
+ */
+export function findFirstSimilarPlanKey( planKey, diff ) {
+	return findSimilarPlansKeys( planKey, diff )[ 0 ];
+}
+
+/**
+ * A similar plan is one that has the same `type`, `group`, and `term` as first
+ * argument, except for differences specified in the second argument.
+ *
+ * For example:
+ *
+ * > findSimilarPlansKeys( TYPE_BUSINESS, { term: TERM_BIENNIALLY } );
+ * [PLAN_BUSINESS_2_YEARS]
+ * > findSimilarPlansKeys( TYPE_JETPACK_BUSINESS_MONTHLY, { type: TYPE_ANNUALLY } );
+ * [TYPE_JETPACK_BUSINESS]
+ *
+ * @param {string|object} planKey Source plan to compare to
+ * @param {object} diff Properties that should differ in matched plan. @see planMatches
+ * @return {string[]} Matched plans keys
+ */
+export function findSimilarPlansKeys( planKey, diff = {} ) {
+	const plan = getPlan( planKey );
+	// @TODO: make getPlan() throw an error on failure. This is going to be a larger change with a separate PR.
+	if ( ! plan ) {
+		return [];
+	}
+	return findPlansKeys( {
+		...pick( plan, 'type', 'group', 'term' ),
+		...diff,
+	} );
+}
+
+/**
+ * Finds all keys of plans matching a query
+ *
+ * For example:
+ *
+ * > findPlansKeys( { term: TERM_BIENNIALLY } );
+ * [PLAN_PERSONAL_2_YEARS, PLAN_PREMIUM_2_YEARS, PLAN_BUSINESS_2_YEARS]
+ *
+ * @param {object} query @see planMatches
+ * @return {string[]} Matched plans keys
+ */
+export function findPlansKeys( query = {} ) {
+	const plans = getPlans();
+	return Object.keys( plans ).filter( k => planMatches( plans[ k ], query ) );
+}
+
+/**
+ * Matches plan specified by `planKey` against `query`.
+ * Only compares `type`, `group`, and `term` properties.
+ *
+ * For example:
+ *
+ * > planMatches( TYPE_BUSINESS, { term: TERM_ANNUALLY, group: GROUP_WPCOM, type: TYPE_BUSINESS } );
+ * true
+ *
+ * > planMatches( TYPE_BUSINESS, { term: TERM_BIENNIALLY } );
+ * false
+ *
+ * @param {string|object} planKey Plan to match
+ * @param {object} query Properties that should match
+ * @return {bool} Does `planKey` match?
+ */
+export function planMatches( planKey, query = {} ) {
+	const acceptedKeys = [ 'type', 'group', 'term' ];
+	const unknownKeys = difference( Object.keys( query ), acceptedKeys );
+	if ( unknownKeys.length ) {
+		throw new Error(
+			`planMatches can only match against ${ acceptedKeys.join( ',' ) }, ` +
+				`but unknown keys ${ unknownKeys.join( ',' ) } were passed.`
+		);
```

Comment:
Be careful of throwing errors when the caller has no intention of catching them. If we'd like to make local dev easier, there are alternative methods:
- Perhaps a simple debug statement
- console.error in dev environments only
- static analysis with a custom eslint rule, which runs in githooks before commit

Reply:
I don't like hiding errors, and that's what debug statement or console.error would do. How would we know something bad occurred if there are no signs of it? Other than browsing console of course - but you have to know to do it and in what circumstances.

Static analysis is better, and it could be just enough when combined with assertions. As a side note I think having assertions would bring calypso-wide benefits.

Reply:
Can't say I'm a big fan of assertions. IMO there's a chance that they might cover up underlying problems, such as code being too permissive or badly designed (think sane defaults/fallbacks). If we're talking about less fragile interfaces, I'd be more interested in static typing.

Reply:
Static typing would be great as well. It is just one side of the coin though - think design by contract. Assertions are not designed for improving app architecture and I don't agree that they may cover up for underlying problems. On the contrary, their only purpose is to help uncover problems. Of course it is possible to write assertions checking if wrong assumptions are fulfilled, the same is possible with unit tests and yet there is a great value in using them.

Reply:
Just to be clear - I'm all for static typing :-)